modules/afprog/afprog.c

```diff
@@ -329,6 +382,23 @@ afprogram_dd_exit(pid_t pid, int status, gpointer s)
 }
 
 static gboolean
+afprogram_dd_restore_reload_info(AFProgramDestDriver * const self, GlobalConfig * const cfg)
+{
+  AFProgramReloadInfo * const restored_info = (AFProgramReloadInfo *)cfg_persist_config_fetch(cfg, afprogram_dd_format_persist_name(self));
+
+  if (restored_info)
+    {
+      self->pid    = restored_info->pid;
+      self->writer = restored_info->writer;
+
+      child_manager_register(self->pid, afprogram_dd_exit, log_pipe_ref(&self->super.super.super), (GDestroyNotify)log_pipe_unref);
+      g_free(restored_info);
```

Comment:
The `pid` should be registered as soon as possible, i.e, right after the open of the process, or after the writer restoration.
The `pid` should be unregistered before terminating the destination program, or before the `self` becomes invalid (because this is a parameter of the `afprogram_dd_exit`). These can happen only in the deinitialization phase.

I hope these guidelines makes the process easier to understand, at the moment I do not know a better way to resolve these constraints in the code.